[SeeYangZhi] iP

SeeYangZhi opened the pull request:
# Implement Feature-Rich Asta Chatbot for Task Tracking
> A reader lives a thousand lives before he dies. The man who never reads lives only one. - George R.R. Martin, [A Dance with Dragons](https://www.goodreads.com/work/quotes/2936175) 

---

Commands Available:
```java
todo
deadline
event
list
mark
unmark
delete
```
---
Task List:
- ~~Add JUnit tests~~
- [ ] Add `find` command

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SeeYangZhi committed:
feat: rename duke to asta

SeeYangZhi committed:
feat: echo

SeeYangZhi committed:
feat: add memory and list function

SeeYangZhi committed:
feat: add marking feature

SeeYangZhi committed:
feat: Implement support for ToDos, Deadlines, and Events with task management functionality

SeeYangZhi committed:
feat: add semi automated text cases

SeeYangZhi committed:
feat: add exception handling

SeeYangZhi committed:
feat: implement delete functionalities

SeeYangZhi committed:
Add enum and switch statements

SeeYangZhi committed:
Update .gitignore

SeeYangZhi committed:
Implement persistent data saving

SeeYangZhi committed:
Merge pull request #1 from SeeYangZhi/branch-Level-7

Completed level 7

SeeYangZhi committed:
Add support for parsing and handling dates and times in tasks

Tasks are currently stored with deadline dates as strings, which limits the ability to perform operations on dates and makes it harder to display dates in a user-friendly format.

Handling dates as strings makes it difficult to perform date comparisons, manipulations, and user-friendly formatting. Using a dedicated date type will enhance code clarity and maintainability.

Let's implement support for parsing dates and times using `java.time.LocalDate` and `java.time.LocalDateTime` in task objects. Accepted formats include `yyyy-mm-dd` (e.g., 2019-10-15). Dates will be displayed in a readable format such as "MMM dd yyyy" (e.g., Oct 15 2019).

Using the Java time package allows for more flexible and robust handling of dates and times, facilitating future enhancements such as filtering tasks by date and adding time-based reminders.

SeeYangZhi committed:
Merge pull request #2 from SeeYangZhi/branch-Level-8

Add support for parsing and handling dates and times in tasks

SeeYangZhi committed:
Refactor code to conform to OOP practices

SeeYangZhi committed:
Merge pull request #3 from SeeYangZhi/A-MoreOOP

Refactor code to conform to OOP practices

SeeYangZhi committed:
Refactor to implement packages

SeeYangZhi committed:
Merge branch 'add-gradle-support'

SeeYangZhi committed:
Update build.gradle

SeeYangZhi committed:
Add TaskListTest.java

SeeYangZhi committed:
Add javadoc

SeeYangZhi committed:
Merge pull request #4 from SeeYangZhi/branch-A-JavaDoc

Add javadoc

SeeYangZhi committed:
Refactor code to proper standard

SeeYangZhi committed:
Merge branch 'master' into branch-A-CodingStandard

SeeYangZhi committed:
Merge pull request #5 from SeeYangZhi/branch-A-CodingStandard

Refactor code to proper standard

SeeYangZhi committed:
Implement find tasks

SeeYangZhi committed:
Merge pull request #6 from SeeYangZhi/branch-Level-9

Implement find tasks

tayxuenye commented on src/main/java/Asta/AstaException.java:
```diff
@@ -0,0 +1,91 @@
+        super(message);
+    }
+
+    /**
```
Great that you used JavaDocs for an extensive list of exceptions!

tayxuenye commented on src/main/java/Asta/Ui.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @return A string representing the user's command input.
+     */
+    public String readCommand() {
```
method names are verbs which is good

tayxuenye commented on src/test/java/Asta/task/TaskListTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test public void testMarkTaskAsDone() {
+        try {
```
Good that you used try-catch for exceptions

tayxuenye commented on src/main/java/Asta/task/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+     * @param byDateStr   The date by which the task should be completed.
+     * @throws AstaException If the description is empty or the date format is invalid.
+     */
+    public void addDeadlineTask(String description, String byDateStr) throws AstaException {
```
Good naming for params

tayxuenye reviewed (commented):
Good job!

seanlim commented on src/main/java/Asta/Asta.java:
```diff
@@ -0,0 +1,116 @@
+            Command command = Parser.parse(fullCommand);
+            ui.showLine();
+            try {
+                switch (command) {
```
Nice switch statement! I see that there is a lot of handler logic nested in each case, do you think that it might be a good idea to create dedicated handler methods for each of the commands?

seanlim commented on src/main/java/Asta/AstaException.java:
```diff
@@ -0,0 +1,91 @@
+ * AstaException is a custom exception class used to handle errors specific to the Asta application.
+ * This class provides static methods to throw specific error messages, ensuring consistent error handling throughout the application.
+ */
+public class AstaException extends Exception {
```
Clean exception handling methods!

seanlim commented on src/main/java/Asta/Storage.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    /**
+     * Loads tasks from the specified file. If the file does not exist, throws an exception.
```
Well written JavaDoc header comments!

seanlim commented on src/main/java/Asta/command/Command.java:
```diff
@@ -0,0 +1,53 @@
+ * The Command enum represents the different types of commands that the Asta application can recognize and execute.
+ * Each enum constant corresponds to a specific user command.
+ */
+public enum Command {
```
Nice use of Java enums!

seanlim reviewed (approved):
LGTM

seanlim reviewed (approved):
LGTM!

SeeYangZhi committed:
Update build.gradle

SeeYangZhi committed:
Update Asta.java

SeeYangZhi committed:
Refactor UI class for GUI display

SeeYangZhi committed:
Merge pull request #7 from SeeYangZhi/branch-Level-10

Branch level 10

SeeYangZhi committed:
Refactor according to style check

SeeYangZhi committed:
add assertions

Add assertions for critical assumptions

To improve code reliability and ensure correctness, this commit introduces assertions at key points in the code to verify assumptions that must hold.

By using assertions, we can document and check important conditions such as:
- Valid task index when accessing, marking, or deleting tasks in TaskList.java.
- Ensuring user input is not null or empty when parsing commands in Parser.java.
- Verifying valid dates during task creation in TaskList.java and Storage.java.

These assertions help catch potential bugs early during development and document critical assumptions in the code.

SeeYangZhi committed:
Merge pull request #8 from SeeYangZhi/branch-A-Assertions

add assertions

SeeYangZhi committed:
Improve code quality

SeeYangZhi committed:
Merge pull request #9 from SeeYangZhi/branch-A-CodeQuality

Improve code quality

SeeYangZhi committed:
refactor findTasks with streams implementation

SeeYangZhi committed:
Add github workflow for Java CI

SeeYangZhi committed:
Update TaskListTest.java

SeeYangZhi committed:
Add initial implementation of recurring deadline task

SeeYangZhi committed:
Add title for app

SeeYangZhi committed:
Implement better responsiveness

SeeYangZhi committed:
add Ui.png

SeeYangZhi committed:
Add User Guide

SeeYangZhi committed:
Add more test for tasks functionalities

SeeYangZhi committed:
Add weird contrast color for asta GUI

SeeYangZhi committed:
Update TaskListTest.java

SeeYangZhi committed:
Fix style based on issue #10

SeeYangZhi committed:
Update index.md
